File: src/main/java/at/flauschigesalex/defaultLibrary/time/TimeHandler.java
package at.flauschigesalex.defaultLibrary.time;

import at.flauschigesalex.defaultLibrary.databases.mongo.LibraryMongoInformation;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.concurrent.TimeUnit;

@SuppressWarnings("unused")
public final class TimeHandler implements Cloneable, LibraryMongoInformation {

    private final long epochMillisecond;

    TimeHandler(final long value) {
        this.epochMillisecond = isEpochMillisecond(value) ? value : value * 1000;
    }

    public static boolean isEpochMillisecond(final @NotNull Long value) {
        return value > 100000000000L;
    }

    public static boolean isEpochSecond(final long value) {
        return !isEpochMillisecond(value);
    }

    /**
     * @param value Timestamp in S or MS
     */
    public static TimeHandler handle(final @NotNull Long value) {
        return new TimeHandler(isEpochMillisecond(value) ? value : value * 1000);
    }

    public static TimeHandler handle(final @NotNull Instant instant) {
        return handle(instant.toEpochMilli());
    }

    public static TimeHandler handle(final @NotNull ZonedDateTime zonedDateTime) {
        return handle(zonedDateTime.toInstant());
    }

    /**
     * @see TimeHandler#clone() Cloneable
     * @deprecated
     */
    public static TimeHandler clone(final @NotNull TimeHandler handler) {
        return handle(handler.getInstant());
    }

    public static TimeHandler now() {
        return handle(Instant.now());
    }

    public long getEpochMilli() {
        return epochMillisecond;
    }

    public long getEpochSecond() {
        return getEpochMilli() / 1000;
    }

    public Instant getInstant() {
        return Instant.ofEpochMilli(getEpochMilli());
    }

    public ZonedDateTime getZonedDateTime(@Nullable ZoneId zoneId) {
        if (zoneId == null)
            zoneId = getDefaultZoneId();
        return ZonedDateTime.ofInstant(getInstant(), zoneId);
    }

    public LocalDateTime getLocalDateTime(@Nullable ZoneId zoneId) {
        if (zoneId == null)
            zoneId = getDefaultZoneId();
        return LocalDateTime.ofInstant(getInstant(), zoneId);
    }

    public ZoneId getDefaultZoneId() {
        return ZoneId.systemDefault();
    }

    public @Deprecated(forRemoval = true) TimeHandler add(final @NotNull TimeUnit timeUnit, @Range(from = 1, to = Long.MAX_VALUE) final long length) {
        return modify(timeUnit, length);
    }

    public TimeHandler plus(final @NotNull TimeUnit timeUnit, @Range(from = 1, to = Long.MAX_VALUE) final long length) {
        return modify(timeUnit, length);
    }

    public TimeHandler subtract(final @NotNull TimeUnit timeUnit, @Range(from = 1, to = Long.MAX_VALUE) final long length) {
        return modify(timeUnit, -length);
    }

    TimeHandler modify(final @NotNull TimeUnit timeUnit, final long value) {
        return handle(timeUnit.toMillis(value));
    }

    public String toString() {
        return String.valueOf(epochMillisecond);
    }

    protected TimeHandler clone() throws CloneNotSupportedException {
        return (TimeHandler) super.clone();
    }
}

File: src/main/java/at/flauschigesalex/defaultLibrary/time/TimeHandler.kt
@file:Suppress("unused")

package at.flauschigesalex.defaultLibrary.time

import at.flauschigesalex.defaultLibrary.time.countdown.Countdown
import at.flauschigesalex.defaultLibrary.time.countdown.CountdownFormat
import at.flauschigesalex.defaultLibrary.time.countdown.CountdownFormat.Companion.default
import java.time.*
import java.util.concurrent.TimeUnit

class TimeHandler private constructor(private val originalMilliSecond: Long) {

    companion object {
        /**
         * Default [ZoneId] used by all TimeHandlers unless explicitly instructed otherwise.
         * @see [TimeHandler.toLocalDateTime]
         * @see [TimeHandler.toZonedDateTime]
         */
        var defaultZoneId: ZoneId = ZoneId.systemDefault()

        /**
         * Creates a TimeHandler with the current time as its value.
         */
        fun now(): TimeHandler = this.ofEpochMilli(System.currentTimeMillis())

        /**
         * Creates a TimeHandler with the provided epoch-millisecond as its value.
         */
        fun ofEpochMilli(millisecond: Long): TimeHandler = TimeHandler(millisecond)
        /**
         * Creates a TimeHandler with the provided epoch-second as its value.
         */
        fun ofEpochSecond(second: Long): TimeHandler = TimeHandler(second * 1000)

        /**
         * Creates a TimeHandler with the provided [instant][Instant] as its value.
         */
        fun ofInstant(instant: Instant): TimeHandler = this.ofEpochMilli(instant.toEpochMilli())
        /**
         * Creates a TimeHandler with the provided [date-time][ZonedDateTime] as its value.
         */
        fun ofZonedDateTime(zdt: ZonedDateTime): TimeHandler = this.ofInstant(zdt.toInstant())

    }

    var epochMillisecond = originalMilliSecond
        private set
    val epochSecond: Long get() = epochMillisecond / 1000

    /**
     * Adds the provided time to its value.
     */
    fun plus(unit: TimeUnit, amount: Long): TimeHandler {
        return change(unit, amount)
    }
    /**
     * Subtracts the provided time from its value.
     */
    fun subtract(unit: TimeUnit, amount: Long): TimeHandler {
        return change(unit, -amount)
    }
    private fun change(unit: TimeUnit, amount: Long): TimeHandler {
        epochMillisecond += unit.toMillis(amount)
        return this
    }

    fun copy(): TimeHandler {
        return TimeHandler(epochMillisecond)
    }
    fun copyOriginal(): TimeHandler {
        return TimeHandler(originalMilliSecond)
    }

    fun toInstant(): Instant {
        return Instant.ofEpochMilli(epochMillisecond)
    }
    fun toDuration(other: TimeHandler): Duration {
        return Duration.ofMillis(this.epochMillisecond - other.epochMillisecond)
    }
    fun toCountdown(other: TimeHandler = now(), formatter: CountdownFormat = default): String {
        return Countdown.create(this.toDuration(other), formatter)
    }

    fun toZonedDateTime(zoneId: ZoneId = defaultZoneId): ZonedDateTime {
        return ZonedDateTime.of(this.toLocalDateTime(zoneId), zoneId)
    }
    fun toLocalDateTime(zoneId: ZoneId = defaultZoneId): LocalDateTime {
        return LocalDateTime.ofInstant(this.toInstant(), zoneId)
    }

    override fun toString(): String {
        return epochMillisecond.toString()
    }
}